Move `Bit` and `Register` to live in Rust space

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

The following commits finally bring `Bits` and `Registers` entirely to Rust by performing a couple of tasks:
- Create native representations of `Bit` (a.k.a. `ShareableQubit` and `ShareableClbit`) and `Register` (a.k.a. `QuantumRegister` and `ClassicalRegister`) and move their python counterparts to be managed by rust PyO3.
- Leverage the usage of native bits and registers in `DAGCircuit` and `CircuitData`.
- Add helper structs `RegisterData` and `BitLocator` to represent the layout of bits/registers in the circuit.

All these changes supersede #13686.

### Details and comments

As we move more of our core data model to Rust, the main drawback that we've still had is needing to use Python space to create bits and registers. Although simple on the outside, the `Bit` and `Register` structures in Python can be quite challenging due to cross referencing that can happen between each type to the other (`Bit._register` <-> `Register[i]`] which is not as easy to replicate in Rust space.

With the following commits we aim to move the representation of `Bits` and `Registers` to Rust in which the behavior is more similar to that of Python and allows us to share instances between circuits, rather than having the circuit have complete ownership of the bits it contains (which is the reason for preferring this over #13686).

### Additions

#### Bit representation:

To represent a `Qubit` or `Clbit` in Rust, we have a couple of identifiers:
- Within the circuit, `Qubit` or `Clbit` which is an index that the circuit can locally refer to when using a bit.
- Globally, we would use one of two types of `Bit`:
    - `ShareableQubit` which represents a global bit.
    - `ShareableClbit` the classical counterpart to the `ShareableQubit`.

The base structure for a `Bit` in Rust, called `BitInfo` describes a global bit as being one of two things:
- `Owned` where the bit is owned by a register.
- `Anonymous` where the bit is independent, containing its own unique ID.

Each of these `Bits` also has an extra property called `BitExtraInfo` that acts as an identifier, and contains any extra information a `Bit` of a specific type should have. In the case of `ShareableQubit` whether the instance is an ancilla or not.

##### Python:
- The python counterparts( `PyBit`, `PyQubit`, `PyClbit` and `PyAncillaQubit` ) have similar behavior and inheritance models as the originals, but should not be subclassed any further as it is no longer supported. See #13841.
- The python counterparts may no longer be comparable via `is()` checks due to conversions in PyO3 not taking on the same addresses.
- Each of the Rust native bits implements `IntoPyObject<'_>` so that they can be successfully converted and extracted from their Python counterparts.

#### Register representation:

To represent a `QuantumRegister` or `ClassicalRegister` in Rust, we can use the rust native counterparts of the same name:
- A rust register is a wrapper to a `RegisterInfo` which classifies them into two categories:
    - `Owning`: which owns its bits, can generate them as needed. All of its bits are `Owned`.
    - `Alias`: A register that acts as a collection of bits independent from their type.
- Registers within a circuit should be stored within a `RegisterData` struct.

##### Python:
- Just as with bits, these structs maintain the same inheritance model and behavior as the originals, but should not be subclassed further. See #13841. They may also not be compared using is(), and the native implementations can be converted to and from Python using the `IntoPyObject<'_> trait.

#### CircuitData
CircuitData has been upgraded to contain registers, this is how it was achieved.
- A helper struct called `RegisterData` was created just to save all of the registers stored in the `CircuitData` instance.
    - This struct contains a mapping between the Register names and instances.
    - Contains a cached `PyDict` to allow for a faster visit path from Python.
- A mapping for qubit instances and its locations has been added in the shape of `BitLocator` it is essentially a wrapper around an `IndexMap` and emulates the behavior of `QuantumCircuit._{cl, qu}bit_indices`.
    - This mapping helps us leverage native types and avoid using Python to fetch the location of a bit.
    - Like `RegisterData`, it also contains a cached `PyDict` to enable faster access from python.
- When processing bit identifiers from Python, we use rust versions of `{q,c}bit_argument_conversions` methods in `QuantumCircuit` which perform mappings with native bit types.

#### DAGCircuit
The DAGCircuit now uses `RegisterData` and `BitLocator` in the same way they are used for `CircuitData`, getting rid of the many `PyDict` objects being stored within. Said additions removed the need of most `py` tokens previously needed to modify the circuit's bits and registers.

### Questions and Comments
Feel free to leave a review/comment addressing any questions you may have. 🚀 

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

raynelfss committed:
Initial: Add `Bit` and `Register` initial structs

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

raynelfss committed:
Add: `AncillaQubit` and `Clbit` python counterparts

raynelfss committed:
Add: `QuantumRegister` and `ClassicalRegister` python counterparts.
- Use `macro_rules!` to create `Bit` and `Registers`.

raynelfss committed:
Add: `AncillaRegister`

raynelfss committed:
Fix: Display methods
- Add `new_owned` method for `BitInfo`.

raynelfss committed:
Fix: Rebalance API to allow for superclasses
- Rebalance currently available api to allow for superclasses `Bit` and `Register` to also live in Rust.
- Remove: `ShareableBit` triat and structs that implement it. Replace them with new structs.

raynelfss committed:
Fix: Temporarily replace python instance, prepare for deletion.
- Add prefix class attribute for python `Register` instances.

raynelfss committed:
Fix: More rebalancing of the API.
- Add `BitExtraInfo` as a soft identifier for `BitInfo`, can be null to identify a `Bit`.
- Add `RegisterInfo::get` method to retrieve the information of a `Bit`.
- Have the rust registers own their counters instead of having them be Python exclusive.
- Make subclassing of `Regster` and `Bit` a bit more effective by helping the subclasses inherit most methods.

raynelfss committed:
Remove: `__getstate__`, `__setstate__` methods from `PyBit`.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit-reg-rust

raynelfss committed:
Fix: Intitial move to using rust native bits in `CircuitData` and `DAGCircuit`.
- Modify `BitData` to accept an extra generic value specifying a "sharable" object type that can also be turned into a `PyObject` to allow for compatibility with `Var`.
- Rename `BitAsKey` to `VarAsKey` as it is currently only used for `Var` instances.
- Modify methods in `CommutationChecker` and `GateDirection` passes to use the newer methods.
- Other tweaks and fixes.

raynelfss committed:
Fix: Implement `FromPyObject<'_>` for `SharableQubit` and `SharableClbit`.
- Remove imports of `QUBIT` and `CLBIT` from `DAGCircuit` and `CircuitData`.

raynelfss committed:
Fix: Discard cache when bits are added to `BitData`
- Discarded old equality and hashing methods for `PyBit` and `PyRegister`.
- Fix `replace_bits` to accept any iterator over `Qubits` and `Clbits` from Python.
- Add `is_empty` methods to Register data, wherever `len` is available.
- Add additional parsing during creation of `Register` to more closely match previous behavior.
- Modify `Bit` tests due to mocking no longer operating correctly.

raynelfss committed:
Fix: `__getitem__` methods in `Register` to preserve data
- Add method `register` to `BitData` to better represent a register reference in an owned bit.

raynelfss committed:
FIx: Streamline constructors for `Register`
- Add missing quotation marks in `Register` repr().

raynelfss committed:
Fix: Invalid `is` check during qpy serialization.
- Bits and registers that live in Rust are not guaranteed to pass the `is` equality check in Python.
- Restore type checking in `DAGCircuit` when adding bits.

raynelfss committed:
Fix: Incorrect serialization of `Bits`
- `Bits` should retain their hash value upon deserialization to avoid incorrect retrievals via hash or indexing. Fixed by implementing correct `__reduce__` method in `PyBit`, `PyClbit`, and `PyQubit`.
- Replace `__getnnewargs__` methods with `__reduce__` for more versatile serialization.
- Extend `SliceOrVec` to include a method that allows a vec with negative indices to correctly iterate based on a provided size.
- Modify `FullAncillaAllocation` to not replace the `QuantumRegister` prefix.
- Add `instances_count` getter attribute to all registers.
- `is` comparisons are no longer guaranteed to work between bits and registers. So some tests have been modified to reflect that.
- When `apply_operation` in the `DAGCircuit` receives a clbit in place of a qubit and viceversa, it will throw a type error, not a Key error. This is handled by PyO3 directly during extraction.

raynelfss committed:
Add: Store registers within `CircuitData`
- Create `RegisterData` struct which, similarly to how `BitData` works, stores the registers and allows to access them by key, in this case by name or index.
- Tweak `CircuitData` and `QuantumCircuit` api to allow for access of registers from within `CircuitData`.
- Add `qubit_indices` and `clbit_indices` structs to keep track of the locations of bits and registers within the circuit.
- Modify `circuit_to_dag` to obtain the registers directly from `CircuitData`.
- Modify `BlueprintCircuit` to rely on `CircuitData` to access `QuantumRegister` instances.
- Add setters and getters for registers.
- Modify `circuit_to_instruction` to bring over the registers into the definition of the `Instruction`.
- Add `contains` method to `BitData`.
- Add rust native `BitLocations` that can be converted to a `Python` version if needed.

raynelfss committed:
Add: Store native registers in `DAGCircuit`.
- Implemented `RegisterData` within the `DAGCircuit`.
- Modified methods of the `DAGCircuit` to utilize new native structures wherever possible.
- Modify `RegisterData` to not have a `description` attribute, and use an `IndexMap` to preserve insertion order of the registers.
- Use native initializers for Registers throughout the crate.
- Use native bits in initializer methods for `QuantumCircuit` and `ClassicalCircuit` instead of `BitData`.
- Modify additional `is` checks from tests where not needed.
- Modify `test_gate_definitions` to create owning registers when testing for definitions from the `EquivalenceLibrary`.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit-reg-rust

raynelfss committed:
Fix: Rebalance `RegisterData` to not use `IndexMap`.
- Fix `remove` method to work more efficiently and add `remove_registers` for multiple removals.
- Store index mapping between register name and `RegisterIndex<u32>` which is copyable.

raynelfss committed:
Add: `BitLocator` to keep a cached python version of bit indices.

raynelfss committed:
Fix: Oxideze `bit_argument_conversion`
- Now that the circuit contains all the available `Qubit` and `Clbit` information, this function can live entirely in Rust.

raynelfss committed:
Fix: Cleanse `DAGCircuit` from having `py` tokens when operating on `Bits` and `Registers`.
- Properly format code and skip certain buggy format checks.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into bit-reg-rust

raynelfss committed:
Docs: Add documentation to python classes and remove old code.

raynelfss committed:
Docs: Add release note

raynelfss committed:
Lint: Fix formatting issues.

mtreinish commented on releasenotes/notes/add-bit-register-rust-979dd113aefb3563.yaml:
```diff
@@ -0,0 +1,23 @@
+  - The :meth:`.QuantumCircuit._qbit_argument_conversion` and :meth:`.QuantumCircuit._cbit_argument_conversion`
+    have been moved to Rust to operate with native data.
+other:
+  - |
```
This should be an upgrade note because it's an api change that users will need to potentially update their usage for on upgrading.

```suggestion
  - |
```

jakelishman replied:
Done in 0e669440a.

mtreinish commented on releasenotes/notes/add-bit-register-rust-979dd113aefb3563.yaml:
```diff
@@ -0,0 +1,23 @@
+    attributes of the :class:`.QuantumCircuit` class have been moved to live in 
+    rust space and store rust native data.
+  - The :meth:`.QuantumCircuit._qbit_argument_conversion` and :meth:`.QuantumCircuit._cbit_argument_conversion`
+    have been moved to Rust to operate with native data.
```
What are the user implications for this? An upgrade note is a potential call for action from end users saying you may need to change something when you upgrade. Just operating in rust space isn't really a user facing change. Looking at the test changes it's the note below about the `is` checks, maybe the `TypeError` instead of `DAGCircuitError`.

raynelfss replied:
Hmm, that's right, these functions weren't exposed to the public either way, so I could remove those parts.

jakelishman replied:
Done in 0e669440a.

mtreinish commented on releasenotes/notes/add-bit-register-rust-979dd113aefb3563.yaml:
```diff
@@ -0,0 +1,23 @@
+   
+   * :class:`.Qubit` and :class:`.QuantumRegister`.
+   * :class:`.Clbit` and :class:`.ClassicalRegister`.
+   * :class:`.AncillaQubit` and :class:`.AncillaRegister`.
```
What's the user impact for the feature here? Is it faster, etc? The rust refactor is not really a user facing change because the interface the users consume hasn't changed. We normally reserve feature notes for user facing features.

raynelfss replied:
No user impact whatsoever, so then should I just remove that section and just mention the upgrade parts?

mtreinish replied:
Yeah, if there isn't a user impact I would just remove this from the release notes

jakelishman replied:
Done in 0e669440a.

mtreinish reviewed (commented):
I haven't looked at the code in any depth, but I had some quick questions on the release notes.

mtreinish commented on qiskit/circuit/register.py:
```diff
@@ -17,217 +17,7 @@
-    def __setstate__(self, state):
-        self._name, self._size, self._hash, self._repr, self._bits = state
-        self._bit_indices = None
+Register = circuit.Register
```
Why not just do: `from qiskit._accelerate.circuit import Register` ?

raynelfss replied:
I was following what we do for importing `CircuitInstruction`, no reason other than that

jakelishman replied:
I got rid of the stub files in 508751975.

mtreinish commented on qiskit/circuit/quantumregister.py:
```diff
@@ -15,61 +15,9 @@
-    # Prefix to use for auto naming.
-    prefix = "a"
-    bit_type = AncillaQubit
+Qubit = circuit.Qubit
```
This is kind of a weird pattern, normally you'd only do this if you needed to rename the object. You can just import the objects. If lint complains I would add an `__all__` to say explicitly the module exports the classes.

jakelishman replied:
I got rid of the stub files in 508751975.

mtreinish commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -1436,6 +1405,8 @@ def __deepcopy__(self, memo=None):
         result._data.replace_bits(
             qubits=_copy.deepcopy(self._data.qubits, memo),
             clbits=_copy.deepcopy(self._data.clbits, memo),
+            qregs=_copy.deepcopy(self._data.qregs, memo),
+            cregs=_copy.deepcopy(self._data.cregs, memo),
```
Not for this PR necessarily, but I feel like this should all be handled by `_data.copy(deepcopy=True)` now. With everything living in rust now, it should be easier to just bake this into the `CircuitData.copy()` implementation (the whole `replace_bits()` call not just the register piece added here).

mtreinish commented on qiskit/circuit/library/blueprintcircuit.py:
```diff
@@ -87,7 +97,7 @@ def qregs(self, qregs):
         self._qregs = []
         self._ancillas = []
-        self._qubit_indices = {}
+        # self._qubit_indices = {}
```
```suggestion
```

mtreinish commented on qiskit/circuit/library/blueprintcircuit.py:
```diff
@@ -39,7 +39,7 @@ def __init__(self, *regs, name: str | None = None) -> None:
         self._qregs: list[QuantumRegister] = []
         self._cregs: list[ClassicalRegister] = []
-        self._qubit_indices = {}
+        # self._qubit_indices = {}
```
```suggestion
```

mtreinish commented on crates/qasm3/src/circuit.rs:
```diff
@@ -235,9 +235,8 @@ impl PyCircuitModule {
-            items: qreg
-                .bind(py)
-                .getattr("_bits")?
+            items: PyList::type_object(py)
+                .call1((qreg.clone(),))?
```
Can't you just use the pyo3 constructor for a `PyList` here? Something like:
```rust
items: PyList::new(py, qreg.clone())?.unbind(),
```

raynelfss replied:
Since these still use the Register types from Python, I believe this should be the one way to use it. I can modify it to use the rust types in the future.

raynelfss committed:
FIx: Remove stray comments
- Add dispose method to `BitLocator` and `RegisterData`.
- Make `BitInfo` private to the `circuit` crate.
- Add `cached_raw` getter to `BitLocator`.

mtreinish reviewed (commented):
I was mid-review and saw you pushed new commits. So here are the comments I have so far. I'll update the branch and continue reviewing, but I wanted to leave the comments in case they got lost in the github webui.

raynelfss committed:
Add: `_check_dup` helper in Rust

mtreinish commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -2583,7 +2583,7 @@ impl PySparseObservable {
                 // The only failure case possible here is the identity, because of how we're
                 // constructing the value to convert.
-                let Ok(term) = ::bytemuck::checked::try_cast((*x as u8) << 1 | (*z as u8)) else {
+                let Ok(term) = ::bytemuck::checked::try_cast(((*x as u8) << 1) | (*z as u8)) else {
```
What are these changes for? They look like fixing clippy on the latest version of rust. Can you save them for a separate PR.

jakelishman replied:
Reverted in 0ff56463b.

raynelfss committed:
Addressing more review comments

mtreinish commented on crates/circuit/src/bit_data.rs:
```diff
@@ -68,40 +113,41 @@ impl PartialEq for BitAsKey {
-pub struct BitData<T> {
+pub struct BitData<T, B> {
     /// The public field name (i.e. `qubits` or `clbits`).
     description: String,
```
Is this field still needed? This was added to track the type of the pyobject. But now that we made this generic over `B` this should be redundant and just take up extra memory for something Rust's type system provides us for free now.

jakelishman replied:
There are nice way to get rid of it, but they'd require a bit more effort than is ideal for the PR in its current state, I think - I'll leave it for now and hit it in a follow-up.

jakelishman replied:
I put all the infrastructure in place to do this in 883d82334, but it's not actually done yet.

jakelishman replied:
Gone in 1b653e840.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+}
+
+macro_rules! create_bit_object {
+    ($name:ident, $extra:ty, $extra_exp:expr, $reg:tt) => {
```
Is `$extra` used at all? I thought it looked weird that `ShareableClbit` set this to `()`, I thought it was pointing to an issue with the `ExtraBitInfo` enum, but actually this was just not used here. So I think we can remove it.

jakelishman replied:
Done in 883d82334.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+            )?
+            .into_super()),
+            false => Bound::new(py, (base, bit)),
+        }
```
The variable names read backwards to me and at first I was confused by the inheritance chain being built here. But then I realized bit was the base bit type and base was the qubit object. I think this would be a lot clearer if you did:

```suggestion
        let py_base_bit = PyBit(self.0.clone());
        let py_qubit = PyQubit(self);
        match ancilla {
            true => Ok(Bound::new(
                py,
                PyClassInitializer::from(py_base_bit)
                    .add_subclass(py_qubit.clone())
                    .add_subclass(PyAncillaQubit(py_qubit.clone())),
            )?
            .into_super()),
            false => Bound::new(py, (py_qubit, py_base_bit)),
        }
```

jakelishman replied:
Obsolete in 883d82334.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+                py,
+                PyClassInitializer::from(bit)
+                    .add_subclass(base.clone())
+                    .add_subclass(PyAncillaQubit(base.clone())),
```
Is this clone necessary? It feels like it's one too many since nothing else is using `base` after this.

jakelishman replied:
Only one of the three classes needed to hold any information at all, which is the new state of the world in 883d82334.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
+        Ok(ob.downcast::<PyClbit>()?.borrow().0.clone())
+    }
+}
```
I feel like these to impls could be put in the macro definition to deduplicate the code between `ShareableQubit` and `SharableClbit` since they're the same between the types and just differ in the typing. The only difference is `IntoPyObject` for the owned type because `ShareableQubit` requires the ancilla handling. But you could do that in the macro definition by matching on extra ancilla arg or something.

jakelishman replied:
As of 883d82334, all the impls are now in the macro, except for `IntoPyObject`, which is a bit finnicky because Python has an inheritance structure (but only for `Qubit`) and Rust doesn't.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+                (slf.get_type(), unique_id).into_pyobject(slf.py())?.hash()
+            }
+        }
+    }
```
I don't think it matters here but in the python space version of this object the hash was cached. I don't think it's important here because we should be doing the hashing of the rust space objects where performance matters for doing look-ups based on a shareable bit object. But I do wonder if there is a python space regression caused by putting bits in a dict.

jakelishman replied:
Fixed in 883d82334.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+        let borrow_other = other.borrow();
+        match (&borrow_slf.0, &borrow_other.0) {
+            (BitInfo::Owned { .. }, BitInfo::Owned { .. }) => {
+                slf.repr()?.as_any().eq(other.repr()?)
```
Surely there is a better way to do this without a string comparison?

jakelishman replied:
Fixed in 883d82334.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+        } else {
+            Ok(())
+        };
+        drop(borrowed_mut);
```
Could you avoid needing to do this manually by moving the `borrow_mut()` call inside the if block?

jakelishman replied:
Gone in 883d82334.

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,588 @@
+}
+
+impl PyBit {
+    /// Quickly retrieves the inner `BitData` living in the `Bit`
```
I might make this docstring a bit more explanatory. The real reason this is needed is so you can create this with an `BitExtraInfo` object to correctly handle the `AncillaQubit` type. Since `BitExtraInfo` isn't exposed to Python we need a separate inner constructor, otherwise this could have just been `new()`

jakelishman replied:
Gone in 883d82334.

mtreinish commented on crates/circuit/src/bit_data.rs:
```diff
@@ -17,6 +17,7 @@ use pyo3::prelude::*;
 use std::hash::{Hash, Hasher};
+use std::sync::OnceLock;
 
 /// Private wrapper for Python-side Bit instances that implements
```
The docstring here needs some updates to indicate this type is only for `Python` `Var` objects now instead of bits.

jakelishman replied:
Fixed in 1b653e840.

mtreinish commented on crates/circuit/src/bit_locator.rs:
```diff
@@ -0,0 +1,126 @@
+    types::{IntoPyDict, PyDict},
+};
+
+pub type BitIndexType<B, R> = IndexMap<B, BitLocations<R>>;
```
This name is very confusing personally I would just remove the alias and just use `IndexMap<B, BitLocations<R>, ahash::RandomState>` directly in the struct as that's a lot clearer. But if you need to use a type alias like this I would rename it `BitIndexMap` to make it clear it's a mapping type (but I'd much prefer removing this alias).

Also we should add the `ahash` hasher here if we're using index map to speedup lookups.

jakelishman replied:
Done in 15228f69a.

mtreinish commented on crates/circuit/src/bit_locator.rs:
```diff
@@ -0,0 +1,126 @@
+    pub fn dispose(&mut self) {
+        self.bit_locations.clear();
+        self.cached.take();
+    }
```
I would add docstrings to these methods since they're marked as public. Especially to make it clear what invalidates the python cache and what doesn't.

jakelishman replied:
Done in 15228f69a.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -38,6 +44,13 @@ use smallvec::SmallVec;
 
 import_exception!(qiskit.circuit.exceptions, CircuitError);
 
+type CircuitDataState<'py> = (
```
I'd throw a docstring on this to make it clear what it's for since it's about 150 lines away from the use:

```suggestion
/// A tuple of a `CircuitData`'s internal state used for pickle's `__setstate__()` method
type CircuitDataState<'py> = (
```

jakelishman replied:
Done in 0f8cf6593.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -152,14 +179,80 @@ impl CircuitData {
+    ///     structures.
+    ///
+    /// Returns:
+    ///     dict(:class:`.QuantumRegister`): The current sequence of registered qubits.
```
Isn't this a list?

jakelishman replied:
Done in 0f8cf6593.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -152,14 +179,80 @@ impl CircuitData {
+        Ok(())
+    }
+
+    /// Returns the a list of registered :class:`.QuantumRegisters` instances.
```
I would reword this to

```suggestion
    /// The a list of registered :class:`.QuantumRegisters` instances.
```

since it's a getter so it'll act like an attribute not a method.

jakelishman replied:
Done in 0f8cf6593.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -186,6 +279,45 @@ impl CircuitData {
+    ///     structures.
+    ///
+    /// Returns:
+    ///     dict(:class:`.ClassicalRegister`): The current sequence of registered qubits.
```
Same comments here as for qregs.

jakelishman replied:
Fixed in 0f8cf6593.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -152,14 +179,80 @@ impl CircuitData {
+        self.qubit_indices.cached(py)
+    }
+
+    #[setter("qregs")]
+    fn set_qregs(&mut self, other: Vec<QuantumRegister>) -> PyResult<()> {
+        self.qregs.dispose();
+        for register in other {
+            self.add_qreg(register, true)?;
+        }
+
+        for (index, qubit) in self.qubits.bits().iter().enumerate() {
+            if !self.qubit_indices.contains_key(qubit) {
+                self.qubit_indices
+                    .insert(qubit.clone(), BitLocations::new(index as u32, []));
+            }
+        }
+        Ok(())
```
I understand this was something we could do before, but I wonder if it makes sense to keep this still, it feels like it doesn't fit with our data model around bits. Like what happens if you have one standalone register with 5 owned qubits and you replace `QuantumCircuit.qregs` with a new standalone register with only 2 bits?

jakelishman replied:
I've left it in place for now, but yeah, it's not a great situation.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -474,15 +689,52 @@ impl CircuitData {
-        qubits: Option<&Bound<PyAny>>,
-        clbits: Option<&Bound<PyAny>>,
+        qubits: Option<Bound<PyAny>>,
+        clbits: Option<Bound<PyAny>>,
```
Is there a reason this can't be a vec of `ShareableQubits` now?

jakelishman replied:
Done in 0f8cf659.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -988,19 +1298,30 @@ impl CircuitData {
+        clbits: BitData<Clbit, ShareableClbit>,
         qargs_interner: Interner<[Qubit]>,
         cargs_interner: Interner<[Clbit]>,
+        qregs: RegisterData<QuantumRegister>,
+        cregs: RegisterData<ClassicalRegister>,
+        qubit_indices: BitLocator<ShareableQubit, QuantumRegister>,
+        clbit_indices: BitLocator<ShareableClbit, ClassicalRegister>,
```
I feel like these arguments are kind of cluttering this constructor. I feel like we should be creating the indices lists for the caller and registers should be optional. The intent of this and the other iterator constructors was to have a quick way to build up a `CircuitData` from an iterator of `PackedInstructions` adding these 4 arguments make it feel like we're requiring a lot of pre-work to create a circuit data like this.

It's fine if we want to clean this up in a follow up though since it's easy to adjust these internal APIs after the fact. But I think this is not the ideal state for these constructors.

jakelishman replied:
I've left this for now, but agree we should re-examine in a follow-up.

mtreinish commented on crates/circuit/src/converters.rs:
```diff
@@ -133,7 +124,8 @@ pub fn dag_to_circuit(
         dag.get_global_phase(),
-    )
+    )?;
+    Ok(circuit)
```
Couldn't you just remove the `let` and the `?;` here? It's not a functional difference so there is no point in changing anything unless there are other pushes and you want to change it.

jakelishman replied:
Done in 239370d0b.

mtreinish commented on crates/circuit/src/dot_utils.rs:
```diff
@@ -19,7 +19,7 @@ use std::collections::BTreeMap;
 
 use crate::dag_circuit::{DAGCircuit, Wire};
-use pyo3::prelude::*;
+use pyo3::{prelude::*, IntoPyObjectExt};
```
I'm surprised clippy doesn't complain about this, but I find mixing `*` in a single line this really hard to read, it's easy enough to just add a separate `use statement` in cases like this.

```suggestion
use pyo3::prelude::*;
use pyo3::IntoPyObjectExt;
```

jakelishman replied:
Done in 239370d0b.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -633,19 +721,19 @@ impl DAGCircuit {
-        let qubits_raw = binding.downcast::<PyList>().unwrap();
-        for bit in qubits_raw.iter() {
-            self.qubits.add(py, &bit, false)?;
+        let qubits_raw = binding.extract::<Vec<ShareableQubit>>().unwrap();
```
Isn't this throwing away a potential python space error now. I'd have expected:

```suggestion
        let qubits_raw = binding.extract::<Vec<ShareableQubit>>()?;
```

I think it was `unwrap()`  before to do the fast path with `downcast` and skip the error handling.

jakelishman replied:
Done in 5347aedca.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -633,19 +721,19 @@ impl DAGCircuit {
-        let clbits_raw = binding.downcast::<PyList>().unwrap();
-        for bit in clbits_raw.iter() {
-            self.clbits.add(py, &bit, false)?;
+        let clbits_raw = binding.extract::<Vec<ShareableClbit>>().unwrap();
```
```suggestion
        let clbits_raw = binding.extract::<Vec<ShareableClbit>>()?;
```

jakelishman replied:
Done in 5347aedca.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1017,107 +1105,78 @@ def _format(operand):
+    pub fn add_qreg(&mut self, qreg: QuantumRegister) -> PyResult<()> {
+        // if !qreg.is_instance(imports::QUANTUM_REGISTER.get_bound(py))? {
+        //     return Err(DAGCircuitError::new_err("not a QuantumRegister instance."));
+        // }
```
```suggestion
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1017,107 +1105,78 @@ def _format(operand):
+        //     return Err(DAGCircuitError::new_err(
+        //         "not a ClassicalRegister instance.",
+        //     ));
+        // }
```
```suggestion
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1183,21 +1248,8 @@ def _format(operand):
-
-        let bit_iter = match self.clbits.map_bits(clbits.iter()) {
+    fn remove_clbits(&mut self, py: Python, clbits: Vec<ShareableClbit>) -> PyResult<()> {
+        let bit_iter = match self.clbits.map_bits(clbits.iter().cloned()) {
```
I'm a bit surprised we need to clone here, since we're passed an owned `Vec` and nothing is touching it after this. Wouldn't `clbits.into_iter()` work?

jakelishman replied:
Unfortunately it's used in the error message a little below.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1390,21 +1424,8 @@ def _format(operand):
-
-        let bit_iter = match self.qubits.map_bits(qubits.iter()) {
+    fn remove_qubits(&mut self, py: Python, qubits: Vec<ShareableQubit>) -> PyResult<()> {
+        let bit_iter = match self.qubits.map_bits(qubits.iter().cloned()) {
```
Same question here about the `cloned()`

jakelishman replied:
Also used in the error message below.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -2925,9 +2965,9 @@ def _format(operand):
-                    if !cargs_set.contains(clbit.clone_ref(py))? {
-                        cargs_list.append(clbit)?;
+                    let clbit = self.clbits.get(*wire).unwrap();
+                    if !cargs_set.contains(clbit.clone())? {
```
A clone on a lookup looks weird here? Can't you just use `&clbit` because that has `IntoPyObject` implemented.

jakelishman replied:
Done in c55ad5c8b.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -2925,9 +2965,9 @@ def _format(operand):
-                    let clbit = &self.clbits.get(*wire).unwrap();
-                    if !cargs_set.contains(clbit.clone_ref(py))? {
-                        cargs_list.append(clbit)?;
+                    let clbit = self.clbits.get(*wire).unwrap();
```
I can't comment on above or where it's used below since the diff on this closure is small and the callers aren't touched. But I'm concerned a bit about the usage of python here. With this PR there shouldn't anything in the data model left that needs to be using qargs and cargs via python like this closure is doing now. We're storing everything as py lists and sets here and I'm not sure why. When bit objects were in rust that kind of made sense since we're mapping the shareable representations but now we have a rust representation of that we can use for mapping.

We still need python for the vars obviously, but it feels like there is more to convert here rather than just adjusting the return typing here to rely on `IntoPyObject`. This might be fine as a follow up optimization, but I'm just concerned about the overhead here.

jakelishman replied:
I agree that this needs redoing, though the data model as changed sufficiently that we can probably redo most of `substitute_node_with_dag` in one fell swoop.  I've run out of time to do it myself now.

mtreinish replied:
I'm thinking we save this for a follow up PR since 2.0.0rc1 is waiting on this. This won't make this any worse than what existed before this PR and we can easily do this for 2.1.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4310,7 +4381,8 @@ def _format(operand):
     fn quantum_causal_cone(&self, py: Python, qubit: &Bound<PyAny>) -> PyResult<Py<PySet>> {
         // Retrieve the output node from the qubit
-        let output_qubit = self.qubits.find(qubit).ok_or_else(|| {
+        let qubit_nat: ShareableQubit = qubit.extract()?;
```
What does `nat` mean here?

jakelishman replied:
I think it's short for "native".

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4382,7 +4454,7 @@ def _format(operand):
 
         let qubits_in_cone_vec: Vec<_> = qubits_in_cone.iter().map(|&&qubit| qubit).collect();
-        let elements = self.qubits.map_indices(&qubits_in_cone_vec[..]);
+        let elements = self.qubits.map_indices(&qubits_in_cone_vec[..]).cloned();
```
Is the `cloned()` needed here? I would assume because we have `IntoPyObject` implemented for `&ShareableQubit` implemented we could just rely on pyo3 to do the copying for us as it's populating the `PySet`.

jakelishman replied:
Fixed in 5347aedca.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -5558,33 +5668,51 @@ impl DAGCircuit {
-            } else if wire.is_instance(imports::CLBIT.get_bound(py))? {
-                NodeType::ClbitIn(self.clbits.find(wire).unwrap())
+            if wire.downcast::<PyQubit>().is_ok() {
+                let qubit: ShareableQubit = wire.extract()?;
```
This feels a bit weird. I feel like `if let Some(py_qubit)` might be better because this will be doing 2 python checks right and we should be able to reuse that check. This applies to all the similar logic for the other types and on output nodes too.

jakelishman replied:
It's `if let Ok(qubit)`, but sure.  Done in 5347aedca.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+    },
+}
+
+// Custom implmentation of Hash to disregard the `IndexMap` used for Alias.
```
```suggestion
// Custom implementation of Hash to disregard the `IndexSet` used for Alias.
```

jakelishman replied:
Obsolete in 883d82334.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+        name: String,
+        bits: Box<IndexSet<BitInfo>>,
+        extra: Option<BitExtraInfo>,
+    },
```
I might want to bikeshed on the names a bit because `Alias` to me reads like an alias for the `Register` not for the bits contained in the register. But I can live with it and the best name I could come up with is using `Owned` and `Referenced`. In qpy we called `Owning` `standalone` but that was a boolean so we didn't have to come up with a name where the bits are owned by something else.

Anyway since I can't come up with a concretely better name I'm fine with these, but I wanted to raise it as a point regardless.

jakelishman replied:
"Alias" came from OQ3 land, and we still map the "aliasing" registers onto OQ3 aliases during export.  I've left these as-is for now.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+impl Hash for RegisterInfo {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match self {
+            RegisterInfo::Owning(owning_register_info) => owning_register_info.hash(state),
```
Did you want to put the discriminant here too?

jakelishman replied:
Obsolete in 883d82334.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+    Owning(Arc<OwningRegisterInfo>),
+    Alias {
+        name: String,
+        bits: Box<IndexSet<BitInfo>>,
```
I would make this:

```suggestion
        bits: Box<IndexSet<BitInfo, ahash::RandomState>>,
```

to speedup lookups.

jakelishman replied:
Obsolete in 883d82334.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+    }
+
+    /// Returns an iterator over the bits within the circuit
+    pub(crate) fn bits(&self) -> Box<dyn ExactSizeIterator<Item = BitInfo> + '_> {
```
The box feels weird here I feel like this could be done without the extra layer of indirection.

jakelishman replied:
You have to `Box` if you've got `dyn Trait` because it's unsized, but this is obsolete in 883d82334 anyway.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+            RegisterInfo::Alias { extra, .. } => match extra {
+                Some(BitExtraInfo::Qubit { is_ancilla }) => *is_ancilla,
+                _ => false,
+            },
```
Is this logic correct? I feel like you could have mixed `AncillaQubit` objects with regular `Qubit` in python without an issue since it was a subclass. So I'm not sure if this is correct since we're only inspecting the first referenced qubit object. I'm thinking either this needs to be `self.0.iter().any(...)`, self.0.iter().all(...)`, oralways return false for `Alias`. Either way maybe this warrants an upgrade release note? I don't think the exact behavior matching python is at all important here because this is the edge of an edge case, but more it is a potential difference we should document.

jakelishman replied:
We check the type safety more thoroughly in 883d82334.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+        let slf_borrow = slf.borrow();
+        let other_borrow = other.borrow();
+        Ok(slf.get_type().eq(other.get_type())?
+            && slf.repr()?.to_string().eq(&other.repr()?.to_string())
```
I'm not sure we should be using a string comparison for this since we have borrowed versions of both object.

jakelishman replied:
Fixed in 883d82334.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+
+impl PyRegister {
+    /// Correctly performs extraction of size and name during creation of a register
+    pub fn inner_parse_new(
```
This method name was confusing at first. It's not parsing anything It's about mapping errors and error messages mostly. I would call it `check_new_inputs()`  or `validate_new_inputs()` to make it clear that this is a validation function not a parser.

jakelishman replied:
Gone in 883d82334.

mtreinish commented on crates/circuit/src/register.rs:
```diff
@@ -0,0 +1,1054 @@
+            }
+        }
+    }
+}
```
This should probably be in `slice.rs` instead of inline here.

jakelishman replied:
This is an extra quirk of `Register`, not standard Python behaviour.  I removed the extra struct in 883d82334 anyway.

mtreinish commented on crates/circuit/src/register_data.rs:
```diff
@@ -0,0 +1,262 @@
+
+use crate::{circuit_data::CircuitError, register::Register};
+
+/// Represents the location in which the register is stored within [RegisterData].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RegisterIndex<R: Register> {
+    index: u32,
+    _marker: PhantomData<R>,
+}
+
+impl<R: Register + Clone> Copy for RegisterIndex<R> {}
```
You can't just derive `Copy` here, I would have expected it to just work without needing to put an explicit marker impl like this?

jakelishman replied:
`PhantomData` is weird about derives sometimes.

mtreinish commented on crates/circuit/src/register_data.rs:
```diff
@@ -0,0 +1,262 @@
+
+#[derive(Debug, Clone)]
+pub struct RegisterData<R: Register> {
+    reg_index: HashMap<String, RegisterIndex<R>>,
+    registers: Vec<R>,
```
Isn't this just an `IndexMap`? I feel like you could do this without needing two fields like this. It would simplify a lot of the implementations here too and you can drop the custom internal index type most likely. If you were having lookup performance issues, it might have just been the hashing algorithm and using `ahash` here should address that.

jakelishman replied:
It reasonably could be, but there's a load of knock-on changes to make where things are exposed as `&[ClassicalRegister]` (which `IndexMap` won't give you), so all those types would need playing with, and I'm out of time.

mtreinish replied:
That's fair, let's leave it for now and save it for a potential future code simplification.

mtreinish reviewed (commented):
I just finished a full pass through the PR. This is looks quite good, thanks for doing this. I'm excited to have this merge. I left several inline comments but nothing seemed like a deal breaker mostly just some code clean-up, questions, and a few naming quibbles. It also isn't that many given the size of this PR.

I am very mildly concerned about some of the implementation details using the new types (or not in several cases), especially where a lot of things seem to clone bits over-eagerly. This should be very inexpensive since it's either an `Arc`, u32, and a bool, or a u64 and a bool, so there is not much overhead for cloning. But in a lot of cases it seems to be clone seems to be before returning one of the new bit objects to python but they all have `IntoPyObject` implemented for reference types so I'm not sure why an explicit clone is necessary I'm concerned there is a larger structural reason for this that I'm missing and might point to something in the data model that isn't quite right because we shouldn't need it if pyo3 will copy the data for us. But this is a very small concern on my part and this is enough of a step forward we can fix that in a follow as all the big pieces are here.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into bit-reg-rust

mtreinish committed:
Fix lint

mtreinish commented:
The test failure reported in https://github.com/Qiskit/qiskit/actions/runs/13653419016/job/38167002314?pr=13860 is an issue with this PR. Doing a little poking locally the `__hash__` and `__eq__` implementations for pybits are letting duplicate bit objects into a set which is what is confusing the transpiler test in the pass and causing the failure.

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into bit-reg-rust

jakelishman committed:
Reword release note

jakelishman committed:
Update HighLevelSynthesis Rust code

jakelishman committed:
Remove stub files related to `Qubit`, `Bit`, `Register`, etc

These were internal-only modules for defining the bit and register
types.  Since those types are now defined by `_accelerate`, there's no
reason to have the stub files.

Pylint makes several complaints about cyclic imports that Python is
perfectly capable of resolving; the choices are either to tell Pylint to
shut up (which we do) or to use overly specific imports in all of
`qiskit.circuit`.

jakelishman committed:
Revert changes to `SparseObservable`

jakelishman committed:
Move register definitions into `bit.rs`

The definition of registers is _really_ tightly coupled to the
definition of bits, and necessitated a lot of extra `pub(crate)`
specifiers to allow them to be separate.  Bringing them into one file
makes it easy to do code-generating macros and keep internal details
internal.

jakelishman committed:
Reimplement `Bit` and `Register` with more type-safe logic

This is a major rewrite of the `Register` system in particular, but
significantly touches `Bit` as well.

This moves the type safety all the way down into the `BitInfo` and
`RegisterInfo` classes, with the exception that Rust space can
_technically_ have minor data incoherence between `AncillaRegister` and
contained qubits, though the internals are private and the constructors
will not allow this.

As a direct result (and as a deliberate aim), `Bit` and `Register` can
no longer be instantiated directly; there's no real concept of the
typeless base classes in Rust, so it takes quite a lot of gymnastics to
keep the old Python form working (where the old Python code wasn't
necessarily the best use of subclassing already).

The dynamic `Iterator` methods are now statically dispatched instead.
Much of the `inner_X` logic is gone; the classes interact a bit more
cleanly with each other.  The bit and register types are constructed
simultaneously by macro, as are their Python-space variants; this
represents how coupled the classes are, and reduces the amount of
`pub(crate)` interfaces needed to split the files.

Changes in the tests are to account for some standard Python errors
changing from `CircuitError` to the more standard `TypeError`, and to
deal with the fact that `Bit` and `Register` cannot be constructed
directly.  Also, some tests inserting insane and undocumented behaviour
like having `QuantumRegister(2.0)` be valid, are gone.

The `__repr__` of `Bit` is now shorter and a bit more legible.

jakelishman committed:
Remove 'description' field from 'BitData'

jakelishman committed:
Fix `CommutationChecker` use of `BitData`

It's plausible that this commit is a minor performance regression
because of the extra `Vec`s, but there's _a bunch_ of improvements to
make in the handling of this pass, so maybe best not to worry about it.

jakelishman committed:
Address review comments in `CircuitData`

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

jakelishman committed:
Address review comments in `BitLocator`

jakelishman committed:
Minor review nits

jakelishman committed:
Fix Rust-space test

jakelishman committed:
Small-scale DAG review comments

jakelishman committed:
Remove a couple of unnecessary clones

jakelishman committed:
Revert "Fix `CommutationChecker` use of `BitData`"

This reverts commit ee1e4468f6feaf3e8b4ca916f43f5c466eec3355.

jakelishman committed:
Hack out the commutation checker

jakelishman commented:
Not sure what's going on in the docs, but the test failure is just a new restriction in `CircuitData::replace_bits` from 0f8cf65, and should be easy to fix (but I'm out of time).

jakelishman commented on crates/circuit/src/bit.rs:
```diff
@@ -103,1514 +98,967 @@ where
-                slf.repr()?.as_any().eq(other.repr()?)
+
+        // "Registers" as a concept are somewhat left over from Python space, so we just make the
+        // class a pyclass directly.  The register objects store an `Arc` internally so that they're
```
This comment is out of date

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13860/commits/b70b8a328acf60753d174c65c3227dc84596e0fb

mtreinish committed:
Fix test failure from replace_bits change

mtreinish committed:
Merge branch 'main' into bit-reg-rust

mtreinish commented on crates/accelerate/src/commutation_checker.rs:
```diff
@@ -83,24 +85,27 @@ static SUPPORTED_ROTATIONS: Lazy<HashMap<&str, Option<OperationRef>>> = Lazy::ne
-    let mut bitdata: BitData<T> = BitData::new(py, "bits".to_string());
+    // Using `VarAsKey` here is a total hack, but the commutation checker could really use a lot
+    // more work than just the typing of things here.  There's a lot more to sort out than just
+    // overuse of Python objects and weird type erasure.
```
```suggestion
    // Using `VarAsKey` here is a total hack, but this is a short-term workaround before a
    // larger refactor of the commutation checker.
```

mtreinish commented on crates/accelerate/src/gate_direction.rs:
```diff
@@ -399,15 +399,12 @@ fn replace_dag(
 //
 // TODO: replace this once we have a Rust version of QuantumRegister
```
Hah I missed this the first time around. maybe we don't need this function anymore. But to minimize the changes in this PR lets leave it as is.

mtreinish committed:
Fix standalone circuit crate build

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1064 @@
+// The trait bounds aren't _strictly_ necessary, but they simplify a lot of later bounds.
+pub trait ShareableBit: Clone + Eq + Hash + Debug {
+    type Subclass: Copy + Eq + Hash + Debug + Default;
+    const DESCRIPTION: &'static str;
```
I'd maybe put a comment here that this is for error messages so we know the human readable bit type name in a generic context.

mtreinish replied:
Oops I forgot to add the comment here. I meant to do it as part of the fix-ups I applied. If I need to push again I'll add it, but otherwise I won't bother.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13860/commits/3348a84ce63e1ddabb9f17582f3842299ca23e46

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1064 @@
+    fn get(&self, index: usize) -> Option<Self::Bit>;
+}
+
+/// Create a (Bit, Register) pair, and the associated Python objects.
```
I know this is an internal macro but Isince there 11 args here I'll add the details to docstring just we know if we come back to this in the future.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/13860/commits/59a9b981bd50ea9ce0fe980cd044d9656ec0fa5f

mtreinish commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1064 @@
+        $reg_counter_name:ident
+    ) => {
+        /// Global counter for the number of anonymous bits created.
+        static $bit_counter_name: AtomicU64 = AtomicU64::new(0);
+        /// Global counter for the number of anonymous register created.
+        static $reg_counter_name: AtomicU32 = AtomicU32::new(0);
```
I'm a bit confused about the typing here. So we allow 2^64 anonymous bits to be created but only 2^32 anonymous registers.  I think this is fine, since creating > 4 billion unnamed registers in a session seems unlikely. I was just surprised to see the mismatch.

jakelishman replied:
Ray's original had the discrepancy, and I just didn't change it, just removed around its declaration a bit. I doubt it'd hurt to make it a u64, and maybe we ought to just to be sure it's never hit.

mtreinish committed:
Add docstring for create_bit_object! macro

mtreinish committed:
Remove old comment

mtreinish commented on crates/circuit/src/bit_locator.rs:
```diff
@@ -0,0 +1,133 @@
+
+    /// Recreate this object from a Python dictionary.
+    pub fn from_py_dict(dict: &Bound<PyDict>) -> PyResult<Self> {
+        let mut locator = Self::new();
+        for (key, value) in dict {
+            locator.insert(key.extract()?, value.extract()?);
+        }
+        Ok(locator)
```
It's not needed for this PR since I'm not 100% sure it's a good idea and it's only a small optimization for an uncommon path. But couldn't we pre-populate `located.cached` with `dict` here? I guess the risk is if it's a shared reference and the input dict is modified. So it's probably not a great idea.

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -186,6 +281,44 @@ impl CircuitData {
+    /// .. warning::
+    ///
+    ///     Do not modify this list yourself.  It will invalidate the :class:`CircuitData` data
+    ///     structures.
```
`CircuitData` isn't public, we wrap it in :class:`.QuantumCircuitData` which is not in the docs apparently but referred to as the return type of `QuantumCircuit.data`

```suggestion
    ///     Do not modify this list yourself.  It will invalidate/corrupt :attr:`.data` for this circuit.
```

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -152,14 +182,79 @@ impl CircuitData {
+    /// .. warning::
+    ///
+    ///     Do not modify this list yourself.  It will invalidate the :class:`CircuitData` data
+    ///     structures.
```
```suggestion
    ///     Do not modify this list yourself.  It will invalidate/corrupt :attr:`.data` for this circuit.
```

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -355,10 +571,10 @@ impl CircuitData {
             }
             for b in self.cargs_interner.get(inst.clbits) {
-                clbits.add(self.clbits.get(*b).unwrap().clone_ref(py))?;
+                clbits.add(self.clbits.get(*b).unwrap().clone())?;
```
These clones aren't needed because `IntoPyObject` does it for us.

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13860/commits/8716c5d201de58053bb208b33d01912eed03c6ea

mtreinish committed:
Remove unnecessary clones from CircuitData::active_bits()

mtreinish commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -518,10 +756,10 @@ impl CircuitData {
                     .unwrap()
                     .unbind(),
-                clbits: PyTuple::new(py, self.clbits.map_indices(clbits))
+                clbits: PyTuple::new(py, self.clbits.map_indices(clbits).cloned())
```
`.cloned()` isn't needed here because it is handled for us by `IntoPyObject`

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/13860/commits/7db3bab7d8249812c130ec3884153af3326fea9d

mtreinish committed:
Remove unnecessary clones from CircuitData::__getitem__()

mtreinish commented on crates/circuit/src/register_data.rs:
```diff
@@ -0,0 +1,260 @@
+    ///
+    /// __**Note:** This operation is performed at `O(n)` times in the worst case.__
+    pub fn remove(&mut self, register: &str) -> Option<R> {
+        self.cached_registers.take();
```
I think it's fine to be conservative here, but I feel like we only need to do this if the if branch is run and we actually mutate the register data.

mtreinish commented on releasenotes/notes/add-bit-register-rust-979dd113aefb3563.yaml:
```diff
@@ -0,0 +1,11 @@
+    :class:`.Bit` and :class:`.Register` (and their subclasses) can no longer be
+    subclassed.  This was never intended to be supported behavior, and doing so
+    would cause unspecified behavior in Qiskit.  It is no longer possible to do
+    this as an implementation detail of the classes.
```
I was thinking we need a release not saying `Bit` can not be instantiated anymore, but we already documented that you shouldn't do that. So a release note probably isn't needed because we weren't supporting it before. Although it might not hurt to be overly pedantic and mention it here.

mtreinish replied:
This was added in: 010cab03a378ca63ca84bb1379bc21aa10b2fc55

mtreinish committed:
Remove unnecessary clones from dag_circuit.rs

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1143,56 +1178,48 @@ impl DAGCircuit {
-        let mut unknown_regs = Vec::new();
-        let self_bound_cregs = self.cregs.bind(py);
+    fn remove_cregs(&mut self, cregs: &Bound<PyTuple>) -> PyResult<()> {
+        // let self_bound_cregs = self.cregs.bind(py);
```
```suggestion
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1143,56 +1178,48 @@ impl DAGCircuit {
+        //         let bit_position = self.clbit_locations.get_mut(&bit).unwrap();
+        //         bit_position.remove_register(&creg, index);
+        //     }
+        // }
```
```suggestion
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -1252,16 +1266,15 @@ impl DAGCircuit {
             }
         }
-        self.remove_qregs(py, &PyTuple::new(py, qregs_to_remove)?)?;
+        self.remove_qregs(&PyTuple::new(py, qregs_to_remove)?)?;
```
Here and in cregs we should remove the pytuple usage since this should be unnecessary now to remove a qreg (or a creg). But we can do this in a follow up too, it won't make anything worse than it is now and this path is uncommon.

mtreinish reviewed (commented):
Thanks for all the working fixing up my earlier review comments @jakelishman, the state of this PR is a lot better now! I have a few small comments inline. 

I think there is definitely going to need to be a series of follow ups after this PR merges to improve our usage of new rust native bits. But given that I think the data model in this PR puts us in a good place I think we can move forward with this and the 2.0.0rc1 and save further improvements for 2.1.0.

mtreinish committed:
Fix modern clippy warnings

mtreinish committed:
Apply suggestions from code review

mtreinish committed:
Adjust drawer register labels

The new repr() implementation for Qubit and QuantumRegister is a better
formatting but it is not valid math mode for drawing. There was an edge
case in the drawer if the circuit has a layout set but the registers in
the virtual qubits of the layout are not present in the layout's
registers it would use the repr for the label. This was previously
something like `Qubit(QuantumRegister(2, "name"))` which while overly
verbose would render. However the intent here is to have a human
readable name so instead this adjusts the logic to just pull the
register name and index from the private register field in the Qubit
object to create that label and if there is no register just use the
index for the layout mapping.

mtreinish reviewed (dismissed):
This LGTM now after making the updates to address my review comments. I also fixed the docs build which was the last blocker in CI.

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+
+            /// Create a new owning register.
+            #[inline]
+            pub fn new_owning(name: String, size: u32) -> Self {
```
I think this should be `pub fn new_owning<S: Into<String>>(name: S, size: u32) -> Self {` or something like that so we can still pass string literals etc. without calling `to_string` on them first.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13860/commits/b3781f5bb1318607a99a8f0ba8c353d613b8ed7f

kevinhartman commented on crates/accelerate/src/gate_direction.rs:
```diff
@@ -399,15 +399,12 @@ fn replace_dag(
-    dag.add_qreg(py, &qreg)?;
+fn add_qreg(dag: &mut DAGCircuit, num_qubits: u32) -> PyResult<Vec<Qubit>> {
+    let qreg = QuantumRegister::new_owning("q".to_string(), num_qubits);
+    dag.add_qreg(qreg.clone())?;
```
It might be nice if we rename `add_qreg` to `py_add_qreg` and then have a native `add_qreg` that returns a reference to the added register, so we can avoid this kind of `clone` more easily.

mtreinish replied:
I definitely agree on this, and almost started to write it myself. But decided it wasn't worth blocking the PR over and were going to have a long tail of internal interface cleanups for rust now that the data model exists.

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+        self.registers.push((register, index))
+    }
+
+    /// Removes a register location on `O(n)`` time, where N is the number of
```
```suggestion
    /// Removes a register location in `O(n)`` time, where N is the number of
```

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+}
+// An internal trait to let `RegisterInfo` manifest full `ShareableBit` instances from `BitInfo`
+// structs without leaking that implemntation detail into the public.
+trait ManifestableBit: ShareableBit {
```
Is this bound necessary? I don't know what this is really for yet, but I would expect to see a trait `Manifestable` with an associated type `Manifest`.

jakelishman replied:
`ShareableBit` is public, and `BitData` uses it. `ManifestableBit` is kind of like the "private extension methods" for it that are only for this module, but you can't control the visibility of trait methods, only the trait. So this is just one way of spelling "these are the private methods".

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+/// Contains the information for a register that owns the bits it contains.
+///
+/// This is separate to the full [RegisterInfo] because owned bits also need to store a
+/// backreference to it.
```
A backreference to what specifically?

mtreinish replied:
It's a backref from the bit object to it's owning register

jakelishman replied:
Owned bits store a backref to the register that owns them, so they can be stand-alone compared without needing to provide a list of all registers the bit _might_ be in. But only owning registers can own bits (ofc) not aliasing ones, so we store the reference to the `OwningRegisterInfo` variant only.

mtreinish replied:
I changed "it"  to "its owning register" in: https://github.com/Qiskit/qiskit/pull/13860/commits/54e78b4795c16f33dbaa0682929057487f34c3b7 to make this more explicit.

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+/// Main representation of the inner properties of a shareable `Bit` object.
+///
+/// This is supplemented by an extra marker type to encode additional subclass information for
+/// communication with Python space.
```
Can we get a bit more of an explanation of what the subclass stuff is? As a fresh pair of eyes, I'm not sure what its purpose is yet at this point in the code.

jakelishman replied:
In short here (agree it could have stood more comment inline): this stuff represents the inheritance tree from Python. `BitInfo` is like the raw `Bit` (though we did away with that object other than as a supertype in this PR) and doesn't represent anything alone. After that you add `Qubit` or `Clbit` in the type system, because those two things are never interchangeable, but we add `Qubit`/`Ancilla` in the bit _value_ system because those two things generally are (though you still occasionally want to tell which it was exactly).

It's largely similar to the `QubitInfo`/`ClbitInfo` system you positted, but this file here _does_ want to act on `Bit` structure generically in order to define the register logic (in `RegisterInfo`), so this allows that. That said, I think in practice I had to do so much in macro boilerplate anyway that it might not have mattered much, but I'm not dead sure how you'd necessarily write the macros with separate structs either.

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+        pub struct $bit_struct(BitInfo<$subclass_ty>);
+        impl $bit_struct {
+            #[inline]
+            fn anonymous_instances() -> &'static AtomicU64 {
```
I'd rather see this as `anonymous_instance_count`.

mtreinish replied:
Renamed in: https://github.com/Qiskit/qiskit/pull/13860/commits/51525ad9f9546a6ee45cd3509af139cfd26de122

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+    }
+}
+
+/// Which subclass a [ShareableClbit] is.
+///
+/// This carries no real data; there's no subclasses of ``Clbit``.  It's mostly used as a data
+/// marker and a way to define traits.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
+pub enum ClbitSubclass {
```
I think I'm not too big a fan of this `Subclass` stuff—it seems a little cumbersome just to identify whether or not a qubit is an ancilla. I'd think we'd instead just have separate Rust structs for `QubitInfo` and `ClbitInfo`, and then have a `bool` inside `QubitInfo` to identify if it's an ancilla.

I might be missing something. Either way, it's something we could refactor later if needed since this isn't part of the public API (yet).

jakelishman replied:
`QubitSubclass` _is_ just a Boolean identifying things, but this way it's got clearer semantics and an easier way to mark which state is the default, I think?

Having separate `QubitInfo` and `ClbitInfo` struts would make it harder to write the register logic, or require us to do ever more stuff inside the macros to code-gen it - I was already a bit unhappy at quite how much needed to go inside the macro here, because it's not as easy to read.

jakelishman replied:
Fwiw, I'm not totally sold on this being the perfect idea, but it was a decent way of letting Rust behave mostly natively with shared structure handled by generics (cf opposed to shared behaviour handled by traits), while having an easy way to map the information to the Python class hierarchy.

mtreinish replied:
Lets save this for a follow-up. Dealing with this one off for ancilla qubits is pretty annoying in general.  I've seen a few attempts at different ways to model this and this is the cleanest so far. 

If we had more forethought on this part of the API we could have deprecated it in early in 1.x for removal in 2.x. This is something we should do in 2.1 almost certainly though.

kevinhartman commented on crates/circuit/src/bit.rs:
```diff
@@ -0,0 +1,1078 @@
+    extends=PyQuantumRegister
+)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PyAncillaRegister;
```
If these aren't intended for use from Rust space, then I think we ought to make them `pub(crate)`, or make them private and add a PyO3 registration function to this module that gets called from `lib.rs`.

Not a hill I want to die on, though.

kevinhartman commented on crates/circuit/src/bit_data.rs:
```diff
@@ -10,98 +10,147 @@
-/// Private wrapper for Python-side Bit instances that implements
-/// [Hash] and [Eq], allowing them to be used in Rust hash-based
-/// sets and maps.
+/// Private wrapper for Python-side objects that implements [Hash] and [Eq], allowing them to be
+/// used in Rust hash-based sets and maps.
```
> super minor: are you able to stop whatever tool it is you have from modifying the line-wrapping when it's not necessary? Just adds a bit more noise in review
> — @jakelishman on one of my PRs

(I see you, I checked the blame 🤣)

jakelishman replied:
in my defence, I had to change some of the words in it too haha

kevinhartman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1626,3 +1975,133 @@ impl<'py> FromPyObject<'py> for AssignParam {
+/// fast lookup), which is assumed to contain the same bits as ``bit_sequence``.
+///
+/// Returns:
+///     List[Bit]: a list of the specified bits from ``bits``.
```
I think this is a Rust-only function, right? But the doc string is written in Python style.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/13860/commits/478cd0171c337e4d69cfd44edbdacb2dd3f7deb2

kevinhartman reviewed (commented):
I've mostly just had a look at the core pieces, and didn't do a complete review of all call-sites into them etc. But this looks solid, nice work!

I don't think any of my comments are blockers, but we may want to address them now or in a follow-up.

mtreinish committed:
Use rustdoc format for bit_argument_conversion()

mtreinish committed:
Make new_owning() generic over type for name

mtreinish committed:
Add comment on DESCRIPTION field in ShareableBit

mtreinish committed:
Add details to BitInfo docstring

mtreinish committed:
Rename anonymous_instances() to anonymous_instance_count()

mtreinish committed:
Cleanup docstring for owning register

mtreinish committed:
Update crates/circuit/src/bit.rs

Co-authored-by: Kevin Hartman <kevin@hart.mn>
